File: github_code/insertion-sort/cpp/52.c
#include "insertionsort.h"
#include <sys/timeb.h>

std::vector<int> InsertionSort::sort(std::vector<int> array){
    timeb start, end;
    ftime(&start); // start timer

    for(unsigned int i=1;i<array.size();i++){
        int j = i;
        while( j>0 && array[j-1] > array[j]){
            int temp = array[j];
            array[j] = array[j-1];
            array[j-1] = temp;
            this->nSwaps += 1;
            this->nComparisons += 1;
            j--;
        }
        this->nComparisons += 1;
    }

    ftime(&end); // end timer
    this->time = (end.time - start.time)*1000;
    this->time=(this->time+end.millitm-start.millitm)/1000;

    return array;
}

File: github_java_sort/7/52.java
package sort;

public class ShellSort {					
	public static void shellSort(int t[]) {
		
		int n = t.length;
		int h = 1;
		while (h <= n / 3) {     
			h = 3 * h + 1;
		}
		while (h > 0) {
			for (int i = h; i < n; i++) {        			
				for (int j = i; j>= h ; j-=h) { 			
					if (t[j] < t[j-h]) {
						int temp = t[j];
						t[j] = t[j-h];
						t[j-h] = temp;
					}
				}
			}			
			h = h / 3;
		}
	}
	
	
	
	
	
	public static void main(String[] args) {
		int tableau[] = { 5, 8, 4, 3, 7, 1, 9, 6, 2, 0 };
		shellSort(tableau);
		display(tableau);
	}
	
	public static void display (int t[]) {
		for (int i = 0; i < t.length; i++) {
			System.out.println(t[i]);
		}
	}
}
